### ISSUE TITLE
[JENKINS-40279] Disabling of Jenkins setup wizard not working as expected

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#10207]
<p>We are deploying and installing Jenkins for customers using a framework of ours, where we are packaging Tomcat, Jenkins, plugins and other maintenance tools to have a full server up and running. </p>

<p>To be able to do this in a controlled manner for the customer, in for example an upgrade situation, we want to disable the setup wizard. </p>

<p><a href="https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties</a><br/>
jenkins.install.runSetupWizard: "<em>Set to false to skip install and upgrade wizards.</em>"</p>

<p>We are setting the system property <b>jenkins.install.runSetupWizard</b> to false to skip both install and upgrade wizards.</p>

<div class="panel" style="border-width: 1px;"><div class="panelContent">
<p>JAVA_OPTS:<br/>
-Xms500m -Xmx500m -XX:MetaspaceSize=128M -XX:MaxMetaspaceSize=1024M  -XX:+HeapDumpOnOutOfMemoryError -XX:+UseCompressedOops -XX:+UseG1GC -XX:MaxGCPauseMillis=400 -XX:GCPauseIntervalMillis=8000 -Djava.awt.headless=true -Dhudson.model.ParametersAction.keepUndefinedParameters=true -Djenkins.install.runSetupWizard=false -javaagent:./dist/ssl/sys-props-agent.jar=props:./dist/ssl/private/syspropsagent.properties -XX:HeapDumpPath=/<path-to-logs>/logs/</p>
</div></div>

<p>But still the setup wizard is appearing when upgrading like below. </p>

<p>Going from:<br/>
Tomcat 7.0.50, Jenkins 1.651.3<br/>
Going to:<br/>
Tomcat 8.5.5, Jenkins 2.19.1, -Djenkins.install.runSetupWizard=false</p>

<p>Upgrading in our terms means: <br/>
/usr/<old-installation-of-framework-incl-tomcat-7-and-jenkins-1.651.3><br/>
/usr/<new-installation-of-framework-incl-tomcat-8-and-jenkins-2.19.1><br/>
/usr/in_production/<link-to-active-installation><br/>
/proj/jenkins_home/ [ JENKINS_HOME is kept separated from installation; plugins, jobs, customer data etc ]</p>

<p>Stop old Tomcat/Jenkins, install new Tomcat/Jenkins in parallel, <b>change symlinks (old->new installation)</b>, start new Tomcat/Jenkins. <br/>
And yes, we understand we might be doing a (parallel) install in your terms, when we perform the above "upgrade". </p>

<p>1) We simply can't understand under which situations this system property is supposed to DISABLE the setup wizard? Is it not supposed to be working when running Jenkins in Tomcat? It IS described to "skip install <b>and</b> upgrade wizards". </p>

<p>As we somehow needed to get this to work we have done extensive googling and trying out several approaches where users seem to have struggling with this as well. One workaround tested was this: <br/>
<a href="https://github.com/geerlingguy/ansible-role-jenkins/issues/50" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/geerlingguy/ansible-role-jenkins/issues/50</a></p>

<p>Where we ONLY are adding the mentioned file with the value of "2.0" (no newline)</p>

<p>Create the file JENKINS_HOME/jenkins.install.UpgradeWizard.state with one line, 2.0<br/>
( echo -n 2.0 > $JENKINS_HOME/jenkins.install.UpgradeWizard.state )<br/>
VOILA, no setup wizard appearing after login to upgraded Jenkins! </p>

<p>2) Is this a supported, recommended way of disabling the wizard? Are there any pitfalls going this way? Or are we simply working around a bug here? </p>

---
<details><summary><i>Originally reported by <a title="Thomas Wikström" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thomas_w">thomas_w</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-40279" target="_blank">Disabling of Jenkins setup wizard not working as expected</a></i></summary>
<i><ul>
<li><b>reported by</b>: thomas_w
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 8
<li><b>watchers</b>: 10
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thomas_w">thomas_w</a>:</i>
<p>No, sorry for the confusion: the Upgrade Wizard (the banner) still appears.<br/>
We aren't either interested in the Setup Wizard being executed, but that is perhaps not the case if the Upgrade Wizard is disabled? </p>

<p>Yes, they both showed up correctly among the System Properties and Environment Variables (JAVA_OPTS).<br/>
(don't have a screen shot right now)</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>:</i>
<p>Seems I misread the source code. It's not a system property but related to the session. Very unusual.</p>

<p>It looks like there's no support way to skip the upgrade wizard entirely. It's not coming up in usual programmatic situations, so apparently the idea was that it's not too difficult to refuse/go through the wizard and remove the plugins afterwards if undesired.</p>

<p>I'll fix the documentation to match reality. Sorry about the confusion.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thomas_w">thomas_w</a>:</i>
<p>Sorry, but we're not only after documentation fixes, Daniel. We think we have two real bugs right now:</p>

<p>1) Disabling the Setup Wizard doesn't work as expected when using system property -Djenkins.install.runSetupWizard=false</p>

<p>Problematic code part we think should be reviewed.</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">https:<span class="code-comment">//github.com/jenkinsci/jenkins/blob/stable-2.19/core/src/main/java/jenkins/install/InstallUtil.java
</span>...
    <span class="code-comment">// tests need <span class="code-keyword">this</span> to be 1.0
</span>    <span class="code-keyword">private</span> <span class="code-keyword">static</span> <span class="code-keyword">final</span> VersionNumber NEW_INSTALL_VERSION = <span class="code-keyword">new</span> VersionNumber(<span class="code-quote">"1.0"</span>);
    <span class="code-keyword">private</span> <span class="code-keyword">static</span> <span class="code-keyword">final</span> VersionNumber FORCE_NEW_INSTALL_VERSION = <span class="code-keyword">new</span> VersionNumber(<span class="code-quote">"0.0"</span>);
...
VersionNumber lastRunVersion = <span class="code-keyword">new</span> VersionNumber(getLastExecVersion());

<span class="code-comment">// Neither the top level config or the lastExecVersionFile have a version
</span><span class="code-comment">// stored in them, which means it's a <span class="code-keyword">new</span> install.
</span><span class="code-keyword">if</span> (FORCE_NEW_INSTALL_VERSION.equals(lastRunVersion) || lastRunVersion.compareTo(NEW_INSTALL_VERSION) == 0) {      <=== When does <span class="code-keyword">this</span> <span class="code-keyword">case</span> normally happen?
    Jenkins j = Jenkins.getInstance();
    
    <span class="code-comment">// Allow <span class="code-keyword">for</span> skipping			<=== If-clause working as expected when reached
</span>    <span class="code-keyword">if</span>(shouldNotRun) {
<span class="code-keyword">try</span> {
    InstallState.INITIAL_SETUP_COMPLETED.initializeState();
    <span class="code-keyword">return</span> j.getInstallState();
} <span class="code-keyword">catch</span> (RuntimeException e) {
    <span class="code-keyword">throw</span> e;
} <span class="code-keyword">catch</span> (Exception e) {
    <span class="code-keyword">throw</span> <span class="code-keyword">new</span> RuntimeException(e);
}
    }
...
}
</pre>
</div></div>

<p>Similar issues where this issue have been touched upon/mentioned/seen:<br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-34007" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-34007</a><br/>
<a href="https://issues.jenkins-ci.org/browse/JENKINS-34035" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-34035</a></p>

<p>2) Suppressing the Upgrade Wizard doesn't work using "session variable" -Djenkins.install.UpgradeWizard.show=false<br/>
Unfortunately this approach number two given to us didn't help us forward. We set it the ordinary way using JAVA_OPTS, right or wrong? <br/>
As mentioned earlier, we can't find the methods accessing this "session variable" being called. </p>

<p>Do you want us to file a separate ticket on this? </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>:</i>
<blockquote><p>When does this case normally happen?</p></blockquote>

<p>Note that this method never returns INITIAL_SECURITY_SETUP without passing the shouldNotRun, so that shouldn't matter.</p>

<p>But the condition you point to evaluates to true iff the .last_exec_version file is empty or contains the version 0.0 OR the JENKINS_HOME/config.xml doesn't exist or contains the version 1.0, indicating the global configuration has never before been saved. The latter is a heuristic to determine whether it's a new install or an existing one, and assumes that all actual Jenkins use involved someone clicking 'Save' on Configure Jenkins at least once (and if not, they're probably not too opposed to seeing a wizard once).</p>

<blockquote><p>Similar issues where this issue have been touched upon/mentioned/seen:</p></blockquote>

<p>One motivated implementing support for the system property you're setting to skip the install wizard, the other is about restoring an accidentally closed upgrade wizard. Neither is really relevant here.</p>

<blockquote><p>2) Suppressing the Upgrade Wizard doesn't work using "session variable" -Djenkins.install.UpgradeWizard.show=false</p></blockquote>

<p>Well, yes, because it doesn't check the system property by that name, so setting it doesn't do anything. This feature simply does not exist, period, and both the documentation and me were wrong. Feel free to file a New Feature request for this.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ringerc">ringerc</a>:</i>
<p>I've also been utterly stumped by this, trying incantations with derivative Docker images of the official <tt>jenkins/jenkins:lts</tt> image like</p>

<p>{{RUN echo $JENKINS_VERSION | tee \<br/>
    /usr/share/jenkins/ref/jenkins.install.UpgradeWizard.state \<br/>
    /usr/share/jenkins/ref/jenkins.install.InstallUtil.lastExecVersion}}</p>

<p>and</p>

<p><tt>ENV JAVA_OPTS "<del>Djenkins.install.runSetupWizard=false ${JAVA_OPTS:</del>}"</tt></p>

<p>But for some reason passing `--env JAVA_OPTS="-Djenkins.install.runSetupWizard=false"` to `docker run` did work. </p>

<p>See also <a href="https://github.com/jenkinsci/docker/issues/310" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/docker/issues/310</a> and <a href="https://github.com/jenkinsci/docker/pull/833" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/docker/pull/833</a></p>
